### DISCUSSION TITLE
Java example of type extensions

### DISCUSSION THREAD

**Discussion by fkorotkov** [graphql-java/graphql-java#2372] [General]
There is an example in the documentation around `type extensions` for a `Human` type on this [page](https://github.com/graphql-java/graphql-java-page/blob/master/content/documentation/v16/schema.md):

```graphql
type Human {
    id: ID!
    name: String!
}

extend type Human implements Character {
    id: ID!
    name: String!
    friends: [Character]
    appearsIn: [Episode]!
}
```

I wonder how to achieve the same in pure Java. I have a `GraphQLSchema` instance already passed to me from another module and I'd like to extend a few types with additional fields. I have hard type trying to find a way to do it in pure Java and any help will be appreciated.

**Comment by bbakerman** [comment]
Schemas in graphql-java are immutable.  However it is possible to
"transform" them into new schemas via a visitor pattern with "change
instructions"

See `graphql.schema.SchemaTransformer`

`graphql.schema.SchemaTransformerTest` is a great place to look for
examples on how to use it.  `graphql.schema.GraphQLTypeVisitor#insertAfter`
is the change command you should use to insert a new field

In your case you want to "add new fields" when you visit a type "Foo"

We need to document this class better because many people are asking this
question.

**Reply by BlinglingGirl** [reply]
I have a similar problem.
I want to insert a new type .In my case I don't care where to insert.

**Reply by bbakerman** [reply]
See my reply your other question
